### ISSUE TITLE
Cannot autowire service "fos_user.mailer": must have a type-hint or be given a value explicitly. Symfony 4

### ISSUE THREAD

**Issue by MustaphaGhlissi** [FriendsOfSymfony/FOSUserBundle#2740]
Hello, 

I am working with a REST API in Symfony 4 with API Platform 

I created a ResettingController from scratch 

But when I call the fos_user.mailer service to send the resetting mail I get the following error : 

Cannot autowire service "fos_user.mailer": argument "$mailer" of method "FOS\UserBundle\Mailer\Mailer::__construct()" must have a type-hint or be given a value explicitly.

Any help please !! 

Thanks before,

**Comment by stof** [comment]
why are you trying to autowire the `fos_user.mailer` service ? Our own service does not have autowiring enabled. Are you overriding this service ?

**Comment by MustaphaGhlissi** [comment]
No, I just called the fos_user.mailer service to send an email 

The service is used by a REST request

**Comment by stof** [comment]
Well, the thing is, this error message does not make any sense:

- FOSUserBundle does not rely on autowiring to configure its own services. We configure them explicitly.
- `fos_user.mailer` is not even a service definition, but an alias, and so would not be able to trigger autowiring.

So the *only* way to get such message would be that some code in your project is overriding the `fos_user.mailer` service. Please search for `fos_user.mailer` in your whole project (including the vendor folder), to see which code could be responsible for that.

**Comment by MustaphaGhlissi** [comment]
this is the action that I am trying to implement for requesting password reset 


 /**
     * @Route("/user/resetting/request/{userEmail}", name="user_password_resetting_request")
     * @Method("GET")
     */
    public function resetPasswordRequestAction(Request $request, $userEmail)
    {
        $user = $this->get('fos_user.user_manager')->findUserByEmail($userEmail);
        if (null === $user) {
            return new JsonResponse(['code'=>404, 'message'=>'User not found']);
        }
        if ($user->isPasswordRequestNonExpired($this->container->getParameter('fos_user.resetting.token_ttl'))) {
            return new JsonResponse(['message'=>'Password request already requested']);
        }
        if (null === $user->getConfirmationToken()) {
            /** @var $tokenGenerator \FOS\UserBundle\Util\TokenGeneratorInterface */
            //$tokenGenerator = $this->get('fos_user.util.token_generator');
            $user->setConfirmationToken($this->tokenGenerator->generateToken());
        }
        $this->get('fos_user.mailer')->sendResettingEmailMessage($user);
        $user->setPasswordRequestedAt(new \DateTime());
        $this->get('fos_user.user_manager')->updateUser($user);
        return new JsonResponse(['code'=>200,'message'=>'Password request send successfully']);
    }

The problem is coming from the line :  $this->get('fos_user.mailer')

Error: 

Controller "App\Controller\ResettingController::resetPasswordRequestAction()" requires that you provide a value for the "$mailer" argument. Either the argument is nullable and no null value has been provided, no default value has been provided or because there is a non optional argument after this one.

**Comment by fayange** [comment]
You can add a value directly in "FOS\UserBundle\Mailer\Mailer::__construct()"=> \ Swift_Mailer $ mailer,  it works too but the best way is override mailer

**Comment by sfblaauw** [comment]
May be this @MustaphaGhlissi ?

```
// services.yaml
services:
    FOS\UserBundle\Mailer\MailerInterface: '@fos_user.mailer.twig_swift'
```
